pandas/io/sql.py

```diff
@@ -837,11 +837,17 @@ def _query_iterator(
         self, result, chunksize, columns, coerce_float=True, parse_dates=None
     ):
         """Return generator through chunked result set."""
+        has_read_data = False
         while True:
             data = result.fetchmany(chunksize)
             if not data:
+                if not has_read_data:
+                    yield DataFrame.from_records(
+                        [], columns=columns, coerce_float=coerce_float
+                    )
```

Comment:
What happens if you just yield an empty DataFrame here without keeping track of having read data?

Reply:
If the query does return rows, the resulting generator will have an empty dangling dataframe. Say the chunksize is set to 5, and the result set has 13 rows, the generator will contain:
```
(
   a DataFrame with 5 rows,
   a DataFrame with 5 rows,
   a DataFrame with 3 rows,
   an empty DataFrame
)
```

This is probably not desirable. I agree, keeping track of if we read data isn't very elegant, but I it's the best solution I can come up with.